app/controllers/api/v0/analytics_controller.rb

```diff
@@ -0,0 +1,66 @@
+module Api
+  module V0
+    include Pundit
+
+    class AnalyticsController < ApplicationController
+      def totals
+        user = get_authenticated_user!
+
+        data = if params[:organization_id]
+                 org = Organization.find_by(id: params[:organization_id])
+                 not_found unless org && belongs_to_org?(user, org)
+
+                 AnalyticsService.new(organization: org).totals
+               else
+                 AnalyticsService.new(user: user).totals
+               end
+        render json: data.to_json
+      end
+
+      def historical
+        raise ArgumentError if params[:start].blank?
+
+        user = get_authenticated_user!
+
+        data = if params[:organization_id]
+                 org = Organization.find_by(id: params[:organization_id])
+                 not_found unless org && belongs_to_org?(user, org)
+
+                 AnalyticsService.new(organization: org, start: params[:start], end: params[:end]).stats_grouped_by_day
+               else
+                 AnalyticsService.new(user: user, start: params[:start], end: params[:end]).stats_grouped_by_day
+               end
+        render json: data.to_json
+      end
+
+      def past_day
+        user = get_authenticated_user!
+
+        data = if params[:organization_id]
+                 org = Organization.find_by(id: params[:organization_id])
+                 not_found unless org && belongs_to_org?(user, org)
+
+                 AnalyticsService.new(organization: org, start: DateTime.current - 1.day).stats_grouped_by_day
+               else
+                 AnalyticsService.new(user: user, start: DateTime.current - 1.day).stats_grouped_by_day
+               end
+        render json: data.to_json
+      end
+
+      private
+
+      def get_authenticated_user!
+        api_token = ApiSecret.find_by(secret: params[:api_token])
+        not_found if api_token.blank?
```

Comment:
should this be an unauthorized error instead of a not found?

Reply:
Yeah I agree with that. I wasn't sure how to render a 401 response though without doing something like the following for each different response:
```ruby
if api_token.blank?
  render status: 401, json: "Error: Invalid API token credentials"
  return
end
```

That seems messy to me, so my quick solution was to raise `not_found`. 🙃

Reply:
Given that 404 results in `{"status": 404, error: "Not Found"}` you should probably return a similar structure. You could add a method "unauthorized" in the base `ApiController` and then call it from there.

Reply:
This can actually be implemented the same way as `not_found`.

```
  def not_found
    raise ActiveRecord::RecordNotFound, "Not Found"
  end
```

```
  def not_authorized
    raise some kind of not authorized error
  end
```

Reply:
AFAIK there's no "unauthorized" error by default in Rails, because it doesn't handle authorization, Devise does.

A workaround to use the exception as an escape pod is to add a custom exception, use `rescue_from` in the controller and use it to render the actual JSON error.

The advantage is that you have a central place to handle all unauthorized errors if you want to measure it somehow